File: nodejs-tools/nodejs-compiler-tools/lexer-xml/src/nextToken.ts
import { skipWhitespace } from "./skipWhitespace.ts";
import {
  EOF_CHARACTER,
  Lexer,
  LexerError,
  Token,
} from "@vighnesh153/lexer-core";
import { readStringLiteral } from "./readStringLiteral.ts";
import {
  isAcceptableIdentifierStart,
  readIdentifier,
} from "./readIdentifier.ts";
import { readComment } from "./readComment.ts";
import { readTextNode } from "./readTextNode.ts";
import { XmlTokenType } from "./tokens.ts";

export function nextToken(lexer: Lexer<XmlTokenType>): Token<XmlTokenType> {
  let t: Token<XmlTokenType>;

  skipWhitespace(lexer);

  const currCh = lexer.inputReader.currentCharacter;
  switch (currCh) {
    case ":": {
      t = createToken(lexer, XmlTokenType.Colon);
      break;
    }
    case "=": {
      t = createToken(lexer, XmlTokenType.Equals);
      break;
    }
    case "<": {
      if (lexer.inputReader.peekCharacter() === "!") {
        const { lineNumber, columnNumber } = lexer.inputReader;
        const commentLiteral = readComment(lexer);
        if (commentLiteral !== null) {
          t = createToken(
            lexer,
            XmlTokenType.CommentLiteral,
            commentLiteral,
            lineNumber,
            columnNumber,
          );
        } else {
          t = createToken(
            lexer,
            XmlTokenType.Illegal,
            `${lexer.inputReader.currentCharacter}`,
          );
        }
      } else {
        t = createToken(lexer, XmlTokenType.LeftAngleBracket);
      }
      break;
    }
    case ">": {
      t = createToken(lexer, XmlTokenType.RightAngleBracket);
      break;
    }
    case "/": {
      t = createToken(lexer, XmlTokenType.ForwardSlash);
      break;
    }
    case "?": {
      t = createToken(lexer, XmlTokenType.QuestionMark);
      break;
    }
    case '"': {
      const { lineNumber, columnNumber } = lexer.inputReader;
      const s = readStringLiteral(lexer);
      t = createToken(
        lexer,
        XmlTokenType.StringLiteral,
        s,
        lineNumber,
        columnNumber,
      );
      break;
    }
    case EOF_CHARACTER: {
      t = createToken(lexer, XmlTokenType.Eof);
      break;
    }
    default: {
      if (
        lexer.currentToken == null ||
        lexer.currentToken.tokenType === XmlTokenType.RightAngleBracket
      ) {
        const { lineNumber, columnNumber } = lexer.inputReader;
        const textNode = readTextNode(lexer);
        t = createToken(
          lexer,
          XmlTokenType.TextNode,
          textNode,
          lineNumber,
          columnNumber,
        );
      } else if (isAcceptableIdentifierStart(currCh)) {
        const { lineNumber, columnNumber } = lexer.inputReader;
        const identifier = readIdentifier(lexer);
        t = createToken(
          lexer,
          XmlTokenType.Identifier,
          identifier,
          lineNumber,
          columnNumber,
        );
      } else {
        lexer.addError(
          new LexerError({
            errorCategory: {
              type: "ILLEGAL_CHARACTER",
              ch: currCh,
            },
            lineNumber: lexer.inputReader.lineNumber,
            columnNumber: lexer.inputReader.columnNumber,
          }),
        );
        t = createToken(lexer, XmlTokenType.Illegal, currCh);
      }
      break;
    }
  }

  lexer.inputReader.readNextCharacter();

  lexer.currentToken = t;
  return t;
}

function createToken(
  lexer: Lexer<XmlTokenType>,
  tokenType: XmlTokenType,
  tokenLiteral: string = tokenType.value,
  lineNumber: number = lexer.inputReader.lineNumber,
  columnNumber: number = lexer.inputReader.columnNumber,
): Token<XmlTokenType> {
  return {
    tokenType: tokenType,
    tokenLiteral: tokenLiteral,
    lineNumber,
    columnNumber,
  };
}

File: tools-jvm/XmlFormatter/src/main/kotlin/dev/vighnesh153/xml/formatter/lexer/nextToken.kt
package dev.vighnesh153.xml.formatter.lexer

fun XmlLexer.nextToken(): Token {
    lateinit var t: Token

    skipWhitespace()

    when (inputReader.currChar) {
        ':' -> {
            t = createToken(TokenType.Colon)
        }

        '=' -> {
            t = createToken(TokenType.Equals)
        }

        '<' -> {
            if (inputReader.peekChar() == '!') {
                val lineNumber = inputReader.lineNumber
                val columnNumber = inputReader.columnNumber
                val commentLiteral = readComment();
                t = if (commentLiteral != null) {
                    createToken(
                        TokenType.CommentLiteral,
                        tokenLiteral = commentLiteral,
                        lineNumber = lineNumber,
                        columnNumber = columnNumber
                    );
                } else {
                    createToken(TokenType.Illegal, tokenLiteral = inputReader.currChar.toString());
                }
            } else {
                t = createToken(TokenType.LeftAngleBracket);
            }
        }

        '>',
        -> {
            t = createToken(TokenType.RightAngleBracket);
        }

        '/' -> {
            t = createToken(TokenType.ForwardSlash);
        }

        '?',
        -> {
            t = createToken(TokenType.QuestionMark);
        }

        '"',
        -> {
            val lineNumber = inputReader.lineNumber
            val columnNumber = inputReader.columnNumber
            val s = readStringLiteral()
            t = createToken(
                TokenType.StringLiteral,
                tokenLiteral = s,
                lineNumber = lineNumber,
                columnNumber = columnNumber
            );
        }

        EOF_CHARACTER -> {
            t = createToken(TokenType.Eof);
        }

        else -> {
            if (currentToken == null || currentToken?.tokenType == TokenType.RightAngleBracket) {
                val lineNumber = inputReader.lineNumber
                val columnNumber = inputReader.columnNumber
                val textNode = readTextNode()
                t = createToken(
                    TokenType.TextNode,
                    tokenLiteral = textNode,
                    lineNumber = lineNumber,
                    columnNumber = columnNumber
                );
            } else if (inputReader.currChar?.isAcceptableIdentifierStart() == true) {
                val lineNumber = inputReader.lineNumber
                val columnNumber = inputReader.columnNumber
                val identifier = readIdentifier()
                t = createToken(
                    TokenType.Identifier,
                    tokenLiteral = identifier,
                    lineNumber = lineNumber,
                    columnNumber = columnNumber
                );
            } else {
                addError(
                    IllegalCharacterLexerError(
                        ch = inputReader.currChar,
                        lineNumber = inputReader.lineNumber,
                        columnNumber = inputReader.columnNumber,
                    )
                );
                t = createToken(TokenType.Illegal, tokenLiteral = inputReader.currChar.toString());
            }
        }
    }

    inputReader.readNextChar()

    currentToken = t
    return t
}

private fun XmlLexer.createToken(
    tokenType: TokenType,
    tokenLiteral: String = tokenType.value,
    lineNumber: Int = inputReader.lineNumber,
    columnNumber: Int = inputReader.columnNumber,
): Token = Token(
    tokenType = tokenType,
    tokenLiteral = tokenLiteral,
    lineNumber = lineNumber,
    columnNumber = columnNumber,
)
